### PR TITLE
Cheetah 2.0

### PR DESCRIPTION
**Pull request by Nashtare**
This PR changes the Cheetah curve, to be now defined over a sextic extension of the so-called Goldilocks primefield `Fp`, of characteristic `p = 2^64 - 2^32 + 1`.

Two notable differences with the previous implementation are that:
- the base primefield `Fp` does not rely on Montgomery representation anymore, but only on regular canonical representation of field elements, which internal value is lying between 0 and p-1 included. This also applies for the coordinates (in `Fp`) of elements of the sextic extension `Fp6`.
- `p` being a full prime (i.e. 64 bits long), there is no wiggle room left within the machine-word holding a field element to store encoding metadata, which requires to use new structs for (un)compressed points, namely CompressedPoint and UncompressedPoint. This yields a total increase from previous implementation of 1 byte in encodings, regardless of the compression flag.

### CODE REVIEW

**Review by 4l0n50 — APPROVED** [review]
Changes on constants and the field operation implementations after changing the characteristic of the field, but also the way of constructing the extension (as I understood) by going directly from Fp to Fp, instead of doing Fp->Fp2->Fp6 as before. The correctness of the new field extension and curve depend more on the tests than what I could add.

File: src/curve.rs [hunk 1]
```diff
@@ -672,12 +525,62 @@ impl_binops_multiplicative_mixed!(AffinePoint, Scalar, ProjectivePoint);
...
+
+    let fa = (&a.c5).mul(&B3.c0).0 as u128;
+    let fb = a.c5.0 as u128;
+    let fb = fb + (fb << 1);
+
+    let c0 = fb * GENERATOR.0 as u128;
+    let c0 = c0 + aa;
+    let c0 = Fp(reduce_u96(c0));
+
+    let c1 = ab + ba;
```
Comment (baumbata) [inline src/curve.rs:556]: Just my OCD: could it be `ba + ab` for consistency with lower lines? Could also have `let c0= aa + c0;` above, but that already breaks the pattern.

Comment (Nashtare) [reply src/curve.rs:556]: That's funny, my OCD logic worked differently there: ordering the ab/ac/ba... alphabetically but when adding something to an existing non-mutable variable I always write the overwritten one on the left. No hard feelings though, so I can change it if you want :)

Comment (baumbata) [reply src/curve.rs:556]: But then doesn't bb come before ca? Anyway, not really an issue. As long as someone's OCD is satisfied :P

File: src/fp6.rs [hunk 2]
```diff
@@ -7,7 +7,7 @@
 // except according to those terms.
 
 //! This module implements arithmetic over the extension field Fp6,
-//! defined with irreducible polynomial v^3 + v + 1.
+//! defined with irreducible polynomial u^6 - u - 1.
```
Comment (baumbata) [inline src/fp6.rs:10]: Is it now u^6 - 7?

File: src/fp6.rs [hunk 3]
```diff
@@ -21,124 +21,136 @@ use serde::de::Visitor;
...
 use crate::utils::square_assign_multi;
 
 use crate::fp::TWO_ADICITY;
 
-// 2^56 root of unity = 3537142827676469596*u + 1182629581807810213
-//                    = 1992083664070583229*u + 2727688745413696580 in Montgomery form
+const BETA: u128 = crate::fp::GENERATOR.0 as u128;
+
+// 2^33 root of unity = 10277652121819048352*u^5 + 9084844568934916810*u^4 + 6141246800624588228*u^3
+//                          + 13627025161062455919*u^2 + 2361889345279789581*u + 3733119278093849254
```
Comment (baumbata) [inline src/fp6.rs:33]: This appears different from the value given below.

File: src/fp6.rs (hunk 2 above)
Comment (Nashtare) [reply src/fp6.rs:10]: oops typo

File: src/fp6.rs (hunk 3 above)
Comment (Nashtare) [reply src/fp6.rs:33]: Forgot to update this, good catch

File: src/curve.rs (hunk 1 above)
Comment (Nashtare) [reply src/curve.rs:556]: Ah I hadn't seen below, fixed this indeed :p

File: src/fp6.rs [hunk 4]
```diff
@@ -1324,18 +1636,17 @@ mod test {
         assert!(bool::from(a.is_zero()));
     }
 
-    #[test]
-    fn test_from_raw_unchecked() {
-        let mut element = Fp6::from_raw_unchecked([4287426845256712189, 0, 0, 0, 0, 0]);
+    // #[test]
+    // fn test_from_raw_unchecked() {
+    //     let mut element = Fp6::from_raw_unchecked([4287426845256712189, 0, 0, 0, 0, 0]);
```
Comment (baumbata) [inline src/fp6.rs:1641]: Do we want to leave this test in but commented out, or remove it?

File: src/fp6.rs [hunk 5]
```diff
@@ -266,76 +310,239 @@ impl Fp6 {
...
-        let bb = (&self.c1).mul(&other.c1);
-        let cc = (&self.c2).mul(&other.c2);
+    /// Computes the multiplication of an Fp6 element with an Fp element
+    pub const fn mul_by_fp(&self, other: &Fp) -> Fp6 {
+        // All helper values computed below are seen as u128 after modular reduction.
+        // This allows for a faster computation of the result coordinates,
+        // by computing all operations in /ZZ, and then finally converting the
+        // values back to Fp through a modular reduction.
+        //
+        // The reduction uses `reduce_u96()` as all final values are less than 96 bits.
```
Comment (baumbata) [inline src/fp6.rs:329]: Does this apply for this function? There aren't really any helper values below, and internally, `Fp::mul` uses `reduce_u128`, right?

**Pull request merged**